Enable adc dma on non gdma ESP32 SoCs

wmrsouza opened the pull request:
This PR enables adc dma operation on:
- esp32
- esp32-s2

wmrsouza committed:
drivers: adc: esp32: enable adc dma on non gdma socs

Enables adc dma on:
- esp32
- esp32-s2

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

raffarost commented on drivers/adc/adc_esp32.c:
```diff
@@ -225,15 +263,38 @@ static int adc_esp32_dma_stop(const struct device *dev)
 
-	adc_digi_pattern_config_t *adc_digi_pattern_config =
-		(adc_digi_pattern_config_t *)pattern_config;
+	adc_digi_pattern_config_t *adc_digi_pattern_config = pattern_config;
```
oh, okay, no I see
sorry

wmrsouza committed:
drivers: adc: esp32: split adc_esp32.c between 2 files

Splits driver source code between adc_esp32.c and adc_esp32_dma.c to
improve maintenance

Signed-off-by: Marcio Ribeiro <marcio.ribeiro@espressif.com>

wmrsouza commented:
> Can we split this into `adc_esp32.c` and `adc_esp32_dma.c`? I would say it would improve a lot maintenance, reading and so.

@sylvioalves
- please take a look at `adc_esp32.c`, `adc_esp32_dma.c`, and  `adc_esp32.h`
- let me know if you would like any additional changes

wmrsouza commented:
> why is this pulling in NXP maintainers?

@decsny I think it's a mistake.
No NXP content is being touched.

sylvioalves commented on drivers/adc/adc_esp32.h:
```diff
@@ -0,0 +1,49 @@
+	adc_hal_dma_ctx_t adc_hal_dma_ctx;
+	uint8_t *dma_buffer;
+	struct k_sem dma_conv_wait_lock;
+#if !SOC_GDMA_SUPPORTED
```
Ok, I see that it is not possible as we want ESP32 and ESP32-S2 to work with DMA support but it does not `SOC_GDMA_SUPPORTED`. Please dismiss it.

sylvioalves commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,484 @@
+#define SPI3_CLK_DEV        ((const struct device *)DEVICE_DT_GET(DT_CLOCKS_CTLR(SPI3_NODE_ID)))
+#define SPI3_CLK_SUBSYS     ((clock_control_subsys_t)DT_CLOCKS_CELL(SPI3_NODE_ID, offset))
+#define SPI3_DMA_CLK_SUBSYS ((clock_control_subsys_t)DT_PROP(SPI3_NODE_ID, dma_clk))
+#endif /* CONFIG_SOC_SERIES_ESP32 */
```
`CONFIG_SOC_SERIES_ESP32-S2`

sylvioalves commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,484 @@
+
+static IRAM_ATTR void adc_esp32_dma_intr_handler(void *arg)
+{
+	if (arg == NULL) {
```
I think `arg == NULL` check can go as the handler is always there.

sylvioalves commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,484 @@
+	struct adc_esp32_data *data = dev->data;
+
+	adc_digi_pattern_config_t *adc_digi_pattern_config = pattern_config;
+	const uint32_t unit_atten_uninit = 999;
```
This number `999` sounds magic number, which is not. I suggested making this global with
`#define UNIT_ATTEN_UNINIT UINT32_MAX` for example.

sylvioalves commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,484 @@
+
+	adc_digi_pattern_config_t *adc_digi_pattern_config = pattern_config;
+	const uint32_t unit_atten_uninit = 999;
+	uint32_t channel_mask = 1, channels_copy = seq->channels;
```
i think `channel_mask` is not need at all as you can handle with only with `channels_copy`, can you check?

wmrsouza replied:
I don't think so since channels_copy is the mask mapping all channels that must be added to adc_digi_pattern_config[] inside the loop and channel_mask only keeps track of the current channel (along with channel_id) for each iteration of the loop

raffarost commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,479 @@
+#ifdef CONFIG_SOC_SERIES_ESP32S2
+	const struct device *spi3_dev = SPI3_DEV;
+
+	if (spi3_dev != NULL) {
```
is this check inverted?

raffarost commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,479 @@
+	}
+
+	if (!device_is_ready(I2S0_CLK_DEV)) {
+		return -ENODEV;
```
should also memory be freed here and other returns in this function? (like done above)
maybe use a common exit for all cases

wmrsouza replied:
nice catch, memory is not been freed after  that esp32 and esp32-s2 denpendent code was added

raffarost commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,479 @@
+		return -EBUSY;
+	}
+
+	unsigned int key = irq_lock();
```
this `lock` session seems quite long
Is it really necessary? `dma_cfg` is local data, so for sure it doesn't need to be inside the lock

wmrsouza replied:
you are absolutely right, thanks

raffarost commented on drivers/adc/adc_esp32.c:
```diff
@@ -734,22 +406,13 @@ static DEVICE_API(adc, api_esp32_driver_api) = {
+#endif /* defined(SOC_GDMA_SUPPORTED) && SOC_GDMA_SUPPORTED */
 
 #define ADC_ESP32_CHECK_CHANNEL_REF(chan)                                                          \
 	BUILD_ASSERT(DT_ENUM_HAS_VALUE(chan, zephyr_reference, adc_ref_internal),                  \
```
I believe GH formatting is the standard (I could be wrong), so I understand probably it's aligned in your editor

uLipe reviewed (commented):
Just a couple of nits, otherwise LGTM

wmrsouza commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,479 @@
+#ifdef CONFIG_SOC_SERIES_ESP32
+	const struct device *i2s0_dev = I2S0_DEV;
+
+	if (i2s0_dev != NULL) {
```
no, it isn't
- adc digital controller on ESP32 requires taking control of i2s0 peripheral to make acquisitions using dma
- this code checks at runtime wether i2s0 has been instantiatd by the i2s driver - when i2s0_dev is not NULL - or is free to be used by the adc driver

sylvioalves commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,443 @@
+		.dev = (void *)SPI_LL_GET_HW(SPI3_HOST),
+		.eof_desc_num = 1,
+		.dma_chan = 3,
+		.eof_step = 1,
```
`.eof_step = 1,` is common, can you move down below?

wmrsouza replied:
this line is common for CONFIG_SOC_SERIES_ESP32 and CONFIG_SOC_SERIES_ESP32S2, but not for those socs that SOC_GDMA_SUPPORTED

wmrsouza commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,443 @@
+		return err;
+	}
+
+#if SOC_GDMA_SUPPORTED
```
I aggree

wmrsouza commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,443 @@
+	dma_blk.block_size = len;
+	dma_blk.dest_address = (uint32_t)buf;
+
+	unsigned int key = irq_lock();
```
fair enough

sylvioalves commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,435 @@
+
+#endif /* SOC_GDMA_SUPPORTED */
+
+#define UNIT_ATTEN_UNINIT UINT32_MAX
```
This could go in the header area instead.

sylvioalves commented on drivers/adc/adc_esp32_dma.c:
```diff
@@ -0,0 +1,435 @@
+		.eof_desc_num = 1,
+		.eof_step = 1,
+#endif /* defined(CONFIG_SOC_SERIES_ESP32) */
+		.eof_num = number_of_adc_samples};
```
What if you break this down?
```c
		.eof_num = number_of_adc_samples,
	};
```
